### PR TITLE
ARROW-6871: [Java] Enhance TransferPair related parameters check and tests

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-6871](https://issues.apache.org/jira/browse/ARROW-6871).

TransferPair related param checks in different classes have potential problems:

i. splitAndTansfer has no indices check in classes like VarcharVector
ii. splitAndTranser indices check in classes like UnionVector is not correct (Preconditions.checkArgument(startIndex + length <= valueCount)), should check params separately.
iii. should add more UT to cover corner cases.

### CODE REVIEW

**Comment by tianchen92** [comment]
cc @siddharthteotia

**Comment by tianchen92** [comment]
also cc @praveenbingo appreciate if you have time to take a look.

File: java/vector/src/main/codegen/templates/UnionVector.java [hunk 1]
```diff
@@ -401,6 +401,8 @@ public TransferPair makeTransferPair(ValueVector target) {
   @Override
   public void copyFrom(int inIndex, int outIndex, ValueVector from) {
     Preconditions.checkArgument(this.getMinorType() == from.getMinorType());
+    Preconditions.checkArgument(inIndex >= 0 && inIndex < from.getValueCount(),
+        "Invalid inIndex: %s", inIndex);
```
Comment (liyafan82) [inline java/vector/src/main/codegen/templates/UnionVector.java]: I am not sure if it is a good idea to add check here. As this method can be performance critical, we decided to add minimum checks here, to compare minor types. 
Please see the discussion here: https://github.com/apache/arrow/pull/5019

Comment (tianchen92) [reply java/vector/src/main/codegen/templates/UnionVector.java]: Thanks for clarify, I reverted changes in this methods.

**Comment by emkornfield** [comment]
@tianchen92 did you add the benchmark you mentioned in the JIRA?

**Comment by tianchen92** [comment]
> @tianchen92 did you add the benchmark you mentioned in the JIRA?

I added a benchmark for copyValueSafe and shows no regression, however, due to https://github.com/apache/arrow/pull/5645#discussion_r347703302, I finally revert the check in copyValueSafe API.

**Comment by tianchen92** [comment]
I add a benchmark and shows no regression.

Before:
Benchmark                                             Mode  Cnt   Score   Error  Units
TransferPairBenchmarks.splitAndTransferIntVector      avgt    5   1.801 ± 0.119  us/op
TransferPairBenchmarks.splitAndTransferVarcharVector  avgt    5  12.794 ± 0.812  us/op

After:
Benchmark                                             Mode  Cnt   Score   Error  Units
TransferPairBenchmarks.splitAndTransferIntVector      avgt    5   1.800 ± 0.102  us/op
TransferPairBenchmarks.splitAndTransferVarcharVector  avgt    5  12.570 ± 1.339  us/op

**Comment by emkornfield** [comment]
+1 thanks.

**Pull request closed without merging**